test/functional/fixtures/fake-lsp-server.lua

```diff
@@ -927,6 +927,65 @@ function tests.basic_formatting()
   }
 end
 
+function tests.basic_watchfiles()
+  skeleton {
+    on_init = function()
+      return {
+        capabilities = {
+          documentFormattingProvider = true,
+        }
+      }
+    end;
+    body = function()
+      local root_dir
+      root_dir = os.tmpname()
```

Comment:
I think it makes the most sense to have created the `root_dir` in the client part of the test, but I couldn't find a simple way of sharing its location with the server without hardcoding it. Sharing it from the server to the client like this did seem to work though, but makes it less practical to use [`helpers.tmpname`](https://github.com/nojnhuh/neovim/blob/232fe82eb106ca955fa5dc4815ef3611b1ee744c/test/helpers.lua#L304) which seems preferable over `os.tmpname`.

Overall, this test server is very tightly coupled with the actual test, so any advice to help make this more flexible is much appreciated.

Reply:
I started preferring using a in-process server instead of spawning a separte nvim instance using stdio as it makes the error handling a bit easier and keeps the server logic closer to the test-case instead of splitting it across 2 files.

See https://github.com/neovim/neovim/pull/19998/files#diff-fb2b7119cd068ae643697e57d3dc10e758737dc763b97b91e5c17736f17b8da8 for an example

Reply:
Thanks for the pointer, that way is definitely much easier to reason about. Just pushed a refactored version of this test using that pattern.

Reply:
The in-process server was recently refactored into a common helpers file you could use directly.